File: node-root/snake-ai/SnakeBrain.ts
import Utils from "snake-game/Utils";
import { Direction } from "snake-game/typing";
import CalcUtils from "./CalcUtils";
import type { ActivationFunction } from "./CalcUtils";

export type LayerShape = [number, number];

export interface ProvidedWeightsAndBiases {
  weights: number[][][];
  biases: number[][];
}

export interface Options {
  inputLength: number;
  layerShapes: LayerShape[];
  hiddenLayerActivationFunction: ActivationFunction;
  providedWeightsAndBiases?: ProvidedWeightsAndBiases;
}

export interface ISnakeBrain {
  inputLength: number;
  layerShapes: LayerShape[];
  readonly hiddenLayerActivationFunction: ActivationFunction;
  weights: number[][][];
  biases: number[][];
}

export default class SnakeBrain implements ISnakeBrain {
  public static readonly OUTPUT_LAYER_LENGTH = 4;

  public static readonly actionMap = Object.freeze({
    0: Direction.UP,
    1: Direction.DOWN,
    2: Direction.LEFT,
    3: Direction.RIGHT,
  }) satisfies Record<number, Direction>;

  private static readonly MIN_WEIGHT = -1;
  private static readonly MAX_WEIGHT = 1;
  private static readonly MIN_BIAS = -1;
  private static readonly MAX_BIAS = 1;

  public static crossOverNumber(a: number, b: number): number {
    return Math.random() < 0.5 ? a : b;
  }

  public inputLength: number;
  public layerShapes: LayerShape[];
  public readonly hiddenLayerActivationFunction: ActivationFunction;
  public weights: number[][][];
  public biases: number[][];

  constructor(options: Options) {
    this.inputLength = options.inputLength;
    this.layerShapes = options.layerShapes;
    this.hiddenLayerActivationFunction = options.hiddenLayerActivationFunction;
    if (!this.validateLayerShapes()) throw new Error("Invalid layer shapes");

    if (options.providedWeightsAndBiases) {
      this.weights = Utils.clone3DArr(options.providedWeightsAndBiases.weights);
      this.biases = Utils.clone2DArr(options.providedWeightsAndBiases.biases);
      if (!this.validateWeightsAndBiases()) throw new Error("Invalid provided weight and bias");
    } else {
      this.weights = this.layerShapes.map((x) => this.generateRandomLayerWeight(x));
      this.biases = this.layerShapes.map((x) => this.generateRandomLayerBias(x));
    }
  }

  public toPlainObject(): ISnakeBrain {
    return {
      inputLength: this.inputLength,
      layerShapes: this.layerShapes,
      hiddenLayerActivationFunction: this.hiddenLayerActivationFunction,
      weights: Utils.clone3DArr(this.weights),
      biases: Utils.clone2DArr(this.biases),
    };
  }

  public validateWeightsAndBiases(): boolean {
    if (this.weights.length === 0 || this.biases.length === 0) return false;
    if (this.weights.length !== this.biases.length) return false;
    if (this.weights.length !== this.layerShapes.length) return false;
    if (this.weights[0][0].length !== this.inputLength) return false;
    if (this.weights[this.weights.length - 1].length !== SnakeBrain.OUTPUT_LAYER_LENGTH) return false;
    if (!this.weights.flat(2).every((x) => x >= SnakeBrain.MIN_WEIGHT && x <= SnakeBrain.MAX_WEIGHT)) return false;
    if (!this.biases.flat(2).every((x) => x >= SnakeBrain.MIN_BIAS && x <= SnakeBrain.MAX_BIAS)) return false;

    for (let i = 1; i < this.weights.length; i++) {
      if (this.weights[i].length !== this.layerShapes[i][0]) return false;
      if (this.weights[i][0].length !== this.layerShapes[i][1]) return false;
    }

    for (let i = 0; i < this.biases.length; i++) {
      if (this.biases[i].length !== this.layerShapes[i][0]) return false;
    }

    return true;
  }

  public compute(input: number[]): Direction {
    const output = CalcUtils.computeMultipleLayer(this.weights, input, this.biases, this.hiddenLayerActivationFunction);
    const index = CalcUtils.indexOfMaxValueInArray(output);
    return SnakeBrain.actionMap[index];
  }

  public crossOver(a: SnakeBrain, b: SnakeBrain): void {
    for (let layerIndex = 0; layerIndex < this.weights.length; layerIndex++) {
      const weight = this.weights[layerIndex];
      const bias = this.biases[layerIndex];

      for (let row = 0; row < weight.length; row++) {
        for (let col = 0; col < weight[row].length; col++) {
          weight[row][col] = SnakeBrain.crossOverNumber(a.weights[layerIndex][row][col], b.weights[layerIndex][row][col]);
        }

        bias[row] = SnakeBrain.crossOverNumber(a.biases[layerIndex][row], b.biases[layerIndex][row]);
      }
    }
  }

  public mutate(mutationRate: number, mutationAmount: number): void {
    for (let layerIndex = 0; layerIndex < this.weights.length; layerIndex++) {
      const weight = this.weights[layerIndex];
      const bias = this.biases[layerIndex];

      for (let row = 0; row < weight.length; row++) {
        for (let col = 0; col < weight[row].length; col++) {
          if (Utils.randomBool(mutationRate)) {
            const newValue = weight[row][col] + Utils.randomUniform(-mutationAmount, mutationAmount);
            weight[row][col] = CalcUtils.minmax(SnakeBrain.MIN_WEIGHT, newValue, SnakeBrain.MAX_WEIGHT);
          }
        }

        if (Utils.randomBool(mutationRate)) {
          const newValue = bias[row] + Utils.randomUniform(-mutationAmount, mutationAmount);
          bias[row] = CalcUtils.minmax(SnakeBrain.MIN_BIAS, newValue, SnakeBrain.MAX_BIAS);
        }
      }
    }
  }

  private validateLayerShapes(): boolean {
    const numOfLayer = this.layerShapes.length;

    if (numOfLayer < 1) return false;

    if (this.layerShapes[0][1] !== this.inputLength) return false;

    if (this.layerShapes[numOfLayer - 1][0] !== SnakeBrain.OUTPUT_LAYER_LENGTH) return false;

    for (let i = 1; i < numOfLayer; i++) {
      if (this.layerShapes[i][1] !== this.layerShapes[i - 1][0]) return false;
    }

    return true;
  }

  private generateRandomLayerWeight(layerShape: LayerShape): number[][] {
    const [row, col] = layerShape;
    const template: number[][] = Array(row)
      .fill(null)
      .map(() => Array(col).fill(null));
    return template.map((x) => x.map((_) => Utils.randomUniform(SnakeBrain.MIN_WEIGHT, SnakeBrain.MAX_WEIGHT)));
  }

  private generateRandomLayerBias(layerShape: LayerShape): number[] {
    const [row, _] = layerShape;
    const template: number[] = Array(row).fill(null);
    return template.map((_) => Utils.randomUniform(SnakeBrain.MIN_BIAS, SnakeBrain.MAX_BIAS));
  }
}

File: kotlin-root/gasnake/src/main/kotlin/snake/ai/SnakeBrain.kt
package com.example.snake.ai

import com.example.snake.game.Utils
import com.example.snake.game.typing.Direction
import kotlin.random.Random

typealias LayerShape = Pair<Int, Int>

data class ProvidedWeightsAndBiases(
    val weights: List<List<MutableList<Double>>>,
    val biases: List<MutableList<Double>>
)

data class Options(
    val inputLength: Int,
    val layerShapes: List<LayerShape>,
    val hiddenLayerActivationFunction: ActivationFunction,
    val providedWeightsAndBiases: ProvidedWeightsAndBiases?
)

data class SnakeBrainData(
    val inputLength: Int,
    val layerShapes: List<LayerShape>,
    val hiddenLayerActivationFunction: ActivationFunction,
    val weights: List<List<List<Double>>>,
    val biases: List<List<Double>>
)


class SnakeBrain(private val options: Options) {
    companion object {
        const val OUTPUT_LAYER_LENGTH = 4
        const val MIN_WEIGHT = -1.0
        const val MAX_WEIGHT = 1.0
        const val MIN_BIAS = -1.0
        const val MAX_BIAS = 1.0

        fun crossOverNumber(a: Double, b: Double): Double {
            return if (Random.nextBoolean()) a else b
        }
    }

    val inputLength: Int = options.inputLength
    val layerShapes: List<LayerShape> = options.layerShapes
    val hiddenLayerActivationFunction: ActivationFunction = options.hiddenLayerActivationFunction
    var weights: List<List<MutableList<Double>>>
    var biases: List<MutableList<Double>>

    init {
        if (!validateLayerShapes()) throw IllegalArgumentException("Invalid layer shapes")

        if (options.providedWeightsAndBiases != null) {
            // todo check immutability problem
            weights = options.providedWeightsAndBiases.weights
            biases = options.providedWeightsAndBiases.biases
            if (!validateWeightsAndBiases()) throw IllegalArgumentException("Invalid provided weight and bias")
        } else {
            weights = layerShapes.map { generateRandomLayerWeight(it) }
            biases = layerShapes.map { generateRandomLayerBias(it) }
        }
    }

    fun toPlainObject(): SnakeBrainData = SnakeBrainData(
        inputLength = inputLength,
        layerShapes = layerShapes,
        hiddenLayerActivationFunction = hiddenLayerActivationFunction,
        weights = weights.map { x -> x.map { y -> y.toList() } },
        biases = biases.map { it.toList() }
    )

    fun validateWeightsAndBiases(): Boolean {
        if (weights.isEmpty() || biases.isEmpty()) return false
        if (weights.size != biases.size) return false
        if (weights.size != layerShapes.size) return false
        if (weights[0][0].size != inputLength) return false
        if (weights.last().size != OUTPUT_LAYER_LENGTH) return false
        if (weights.flatten().flatten().any { it < MIN_WEIGHT || it > MAX_WEIGHT }) return false
        if (biases.flatten().any { it < MIN_BIAS || it > MAX_BIAS }) return false

        for (i in 1 until weights.size) {
            if (weights[i].size != layerShapes[i].first) return false
            if (weights[i][0].size != layerShapes[i].second) return false
        }

        for (i in biases.indices) {
            if (biases[i].size != layerShapes[i].first) return false
        }

        return true
    }

    fun compute(input: List<Double>): Direction {
        val output = CalcUtils.computeMultipleLayer(weights, input, biases, hiddenLayerActivationFunction)
        val index = CalcUtils.indexOfMaxValueInArray(output)
        return Direction.inverseDirectionMap(index)
    }

    fun crossOver(a: SnakeBrain, b: SnakeBrain) {
        for (layerIndex in weights.indices) {
            val weight = weights[layerIndex]
            val bias = biases[layerIndex]

            for (row in weight.indices) {
                for (col in weight[row].indices) {
                    weight[row][col] = crossOverNumber(a.weights[layerIndex][row][col], b.weights[layerIndex][row][col])
                }

                bias[row] = crossOverNumber(a.biases[layerIndex][row], b.biases[layerIndex][row])
            }
        }
    }

    fun mutate(mutationRate: Double, mutationAmount: Double) {
        for (layerIndex in weights.indices) {
            val weight = weights[layerIndex]
            val bias = biases[layerIndex]

            for (row in weight.indices) {
                for (col in weight[row].indices) {
                    if (Utils.randomBool(mutationRate)) {
                        val newValue = weight[row][col] + Utils.randomUniform(-mutationAmount, mutationAmount)
                        weight[row][col] = CalcUtils.minmax(MIN_WEIGHT, newValue, MAX_WEIGHT)
                    }
                }

                if (Utils.randomBool(mutationRate)) {
                    val newValue = bias[row] + Utils.randomUniform(-mutationAmount, mutationAmount)
                    bias[row] = CalcUtils.minmax(MIN_BIAS, newValue, MAX_BIAS)
                }
            }
        }
    }

    private fun validateLayerShapes(): Boolean {
        val numOfLayer = layerShapes.size

        if (numOfLayer < 1) return false

        if (layerShapes[0].second != inputLength) return false

        if (layerShapes[numOfLayer - 1].first != OUTPUT_LAYER_LENGTH) return false

        for (i in 1 until numOfLayer) {
            if (layerShapes[i].second != layerShapes[i - 1].first) return false
        }

        return true
    }

    private fun generateRandomLayerWeight(layerShape: LayerShape): List<MutableList<Double>> {
        val (row, col) = layerShape
        return List(row) { MutableList(col) { Random.nextDouble(MIN_WEIGHT, MAX_WEIGHT) } }
    }

    private fun generateRandomLayerBias(layerShape: LayerShape): MutableList<Double> {
        val (row, _) = layerShape
        return MutableList(row) { Random.nextDouble(MIN_BIAS, MAX_BIAS) }
    }
}